File: 动态规划/494.目标和.py
#
# @lc app=leetcode.cn id=494 lang=python
# @lcpr version=30120
#
# [494] 目标和
#


# @lcpr-template-start

# @lcpr-template-end
# @lc code=start
class Solution(object):
    def findTargetSumWays(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        dict = {0:1}    # 长度至少为1，所以可以这么初始化。如果长度可以为0，这么初始化，如果nums中没有0且target为0，会出错返回1

        for num in nums:
            dict_new = {}
            for key in dict.keys(): # dict.keys()访问所有key
                if dict_new.get(key+num) is  None:      # dict.get()是表达式，不是赋值目标.不存在则返回None
                    dict_new[key+num]=0                 # 如果不在dict中，初始化为0
                dict_new[key+num]+=dict.get(key)        # 如果需要修改和赋值使用下表索引

                if dict_new.get(key-num) is  None:      # 厘清key和value的含义关系
                    dict_new[key-num]=0
                dict_new[key-num]+=dict.get(key)
            dict = dict_new
            # print(dict)   # python中字典的打印非常方版
        return 0 if dict.get(target) is None else dict.get(target)  # 不能使用?:，使用if else
# @lc code=end



#
# @lcpr case=start
# [1,1,1,1,1]\n3\n
# @lcpr case=end

# @lcpr case=start
# [1]\n1\n
# @lcpr case=end

#


if __name__ == "__main__":
    a = [1]
    b = int(input())
    print(Solution().findTargetSumWays(a,b))

File: 动态规划/背包问题/494.目标和.cpp
/*
 * @lc app=leetcode.cn id=494 lang=cpp
 * @lcpr version=30121
 *
 * [494] 目标和
 */


// @lcpr-template-start
using namespace std;
#include <algorithm>
#include <array>
#include <bitset>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <stack>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
// @lcpr-template-end
// @lc code=start
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        // T=A-S => 2S=A+S-T 
        int total=-target;
        for(auto &num: nums) total+=num; // 注意下面使用total

        if(total<0) return false; // total至少非负！否则会出现
        // Line 1037: Char 9: runtime error: reference binding to null pointer of type 'int' (stl_vector.h)
        // 表示数组访问出界

        vector<int> dp(total+1, 0); //总表达式的数目，初始化为0!
        dp[0]=1;

        for(auto &num:nums){
            for(int j=total;j>=2*num;j--){
                dp[j]=dp[j]+dp[j- 2*num];
            }
        }

        return dp[total];
    }
};
// @lc code=end



/*
// @lcpr case=start
// [1]\n2\n
// @lcpr case=end

// @lcpr case=start
// [1]\n1\n
// @lcpr case=end

 */

